### ISSUE TITLE
[HELP] Perform backpropagation separately for the two losses and extract the corresponding gradients.

### ISSUE THREAD

**Issue by 1170300714** [microsoft/DeepSpeed#7352]
Hello, I currently have a requirement: during a single-time forward pass, I compute two losses, and I would like to backpropagate each loss **separately** to obtain their **respective** gradients. Since I made the modifications within a complex project codebase and cannot paste the original code, the following pseudocode illustrates the core logic: 

```python
#step1:  forward, and calculate two losses
output = model(input)
loss1 = loss1_func(output, target_1)
loss2 = loss2_func(output, target_2)

#step2:  loss1 backward and extract its grad
model.zero_grad()
accelerator.backward(loss1,  retain_graph=True)
grad_loss1 = safe_get_full_grad(model)

#step3:  loss2 backward and extract its grad
model.zero_grad()
accelerator.backward(loss2)
grad_loss2 = safe_get_full_grad(model)
```
However, when I run the code with **ZeRO-1** and **gradient_accumulation_steps=1**, I find that the second call to model.zero_grad() (the first line of step3 ) does not take effect — that is, even though this line is executed, the gradients from loss1 still remain in the model. Moreover, the gradients from loss2 cannot overwrite or accumulate onto the existing gradients. I**n other words, once the backward pass for loss1 is completed, its gradients are forcibly retained in the model and cannot be cleared or overwritten.**

I further found that when I set gradient_accumulation_steps=2, the gradients for the second loss can be backpropagated correctly. However, in my project, gradient_accumulation_steps must be strictly set to 1. Therefore, is there any way to successfully extract the gradients corresponding to both losses when gradient_accumulation_steps=1?

Thanks a lot!

**Comment by tjruwase** [comment]
@1170300714 , this is an interesting problem. Would it be accurate to summarize the problem as `model.zero_grad()` is not working correctly since it clears gradients as expected for `gradient_accumulation_step=2` but not for `gradient_accumulation_step=1`?  

Is it possible to test with `ZeRO=0` (i.e., set your zero stage to 0 which becomes DDP)? Does the problem show in that case?

It would be very helpful if you could provide a standalone small repro. It seems to me that your pseudo-code is almost that and what is needed is definitions for `loss*_func`, `target_*` and `input`. Could you help with that extra step of providing a repro?  
Thanks

**Comment by 1170300714** [comment]
收到，感谢您的来件

**Comment by tjruwase** [comment]
> 收到，感谢您的来件

@1170300714 , I can't understand your reply. Is this still an issue?

**Comment by therealnaveenkamal** [comment]
@tjruwase @1170300714 

This is an interesting observation. I've written a repro for this issue. 

Script:

```
import torch
import torch.nn as nn
import deepspeed
import argparse
import os
from deepspeed.utils import safe_get_full_grad
from deepspeed.accelerator import get_accelerator

class SimpleModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear1 = nn.Linear(10, 20)
        self.linear2 = nn.Linear(20, 10)
        
    def forward(self, x):
        x = torch.relu(self.linear1(x))
        return self.linear2(x)

def loss2_func(output, target):
    return nn.MSELoss()(output, target)

def loss1_func(output, target):
    return nn.MSELoss()(output, target)

def get_full_grads(model_engine):
    grads = []
    for param in model_engine.parameters():
        grad = safe_get_full_grad(param)
        if grad is not None:
            grads.append(grad.clone().flatten())
    
    if not grads:
        print("Warning: No grads!")
        return None
    
    return torch.cat(grads)


model = SimpleModel()
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = model.to(device)

torch.manual_seed(42)
input_data = torch.ones(32, 10, device=device)  # All ones for simplicity
target1 = torch.zeros(32, 10, device=device)    # All zeros for target1
target2 = torch.ones(32, 10, device=device)     # All ones for target2

# Initialize DeepSpeed with ZeRO-1 configuration
model_engine, _, _, _ = deepspeed.initialize(
    model=model,
    model_parameters=model.parameters(),
    config={
        "train_batch_size": 32,
        "gradient_accumulation_steps": 1,
        "optimizer": {
            "type": "Adam",
            "params": {
                "lr": 1e-3
            }
        },
        "zero_optimization": {
            "stage": 1,
            "overlap_comm": False,
            "contiguous_gradients": True
        },
        "fp16": {
            "enabled": True,
            "loss_scale": 1024,
            "initial_scale_power": 10,
        }
    }
)

if model_engine.fp16_enabled() and get_accelerator().is_fp16_supported():
    input_data = input_data.half()
    target1 = target1.half()
    target2 = target2.half()

# Forward pass
output = model_engine(input_data)
loss1 = loss1_func(output, target1)
loss2 = loss2_func(output, target1)

print("\nComputing gradients for loss1")
# Backward for loss1
model_engine.zero_grad()
model_engine.backward(loss1, retain_graph=True)
grad_loss1 = get_full_grads(model_engine)
print("Gradients after loss1:", grad_loss1.norm().item())


print("\nComputing gradients for loss2")
# Backward for loss2
model_engine.zero_grad()
model_engine.backward(loss2)
grad_loss2 = get_full_grads(model_engine)
print("Gradients after loss2:", grad_loss2.norm().item())
```



Output when `gradient_accumulation_steps = 2`

```
2025-06-17 02:31:17,032] [INFO] [config.py:911:print_user_config]   json = {
    "train_batch_size": 32, 
    "gradient_accumulation_steps": 2, 
    "optimizer": {
        "type": "Adam", 
        "params": {
            "lr": 0.001
        }
    }, 
    "zero_optimization": {
        "stage": 1, 
        "overlap_comm": false, 
        "contiguous_gradients": true
    }, 
    "fp16": {
        "enabled": true, 
        "loss_scale": 1.024000e+03, 
        "initial_scale_power": 10
    }
}

Computing gradients for loss1
Gradients after loss1: 210.875

Computing gradients for loss2
Gradients after loss2: 210.875
```

Output when `gradient_accumulation_steps = 1`

```
2025-06-17 02:33:55,556] [INFO] [config.py:911:print_user_config]   json = {
    "train_batch_size": 32, 
    "gradient_accumulation_steps": 1, 
    "optimizer": {
        "type": "Adam", 
        "params": {
            "lr": 0.001
        }
    }, 
    "zero_optimization": {
        "stage": 1, 
        "overlap_comm": false, 
        "contiguous_gradients": true
    }, 
    "fp16": {
        "enabled": true, 
        "loss_scale": 1.024000e+03, 
        "initial_scale_power": 10
    }
}

Computing gradients for loss1
Gradients after loss1: 228.31080627441406

Computing gradients for loss2
Gradients after loss2: 456.6216125488281

```

So essentially, the loss1 and loss2 should be the same. But here, loss2 gradients are 2xloss1 gradients - which means zero_grad() didn't clear the grads I believe.

**Comment by therealnaveenkamal** [comment]
As per my understanding, in gradient_accumulation_steps = 1, after backward(), DS considers that the grad accumulation is done and is now waiting for model_engine.step(). 

Since the grad accumulation is complete, zero_grad() doesn't work. 

Whereas in gradient_accumulation_steps = 2, the accumulation is still in progress after backward() call and change is still permitted. So grads can be cleared and accumulated again.
